### PR TITLE
Lukeu/ide/eclipse project when merged in buildship

### PR DESCRIPTION
**Pull request by lukeu**
### Context

The general expectation is that Buildship's "Refresh Gradle Project" action acts the same as `gradle eclipse`, but in reality there are holes. This plugs one such hole by ensuring that code declared in `eclipse.project.file.whenMerged` gets executed.

This approach was suggested by @oehme [in Buildship Issue 694](https://github.com/eclipse/buildship/issues/694#issuecomment-383132069)

### Contributor Checklist
- [x] [Review Contribution Guidelines](https://github.com/gradle/gradle/blob/master/.github/CONTRIBUTING.md)
- [x] Make sure that all commits are [signed off](https://git-scm.com/docs/git-commit#git-commit---signoff) to indicate that you agree to the terms of [Developer Certificate of Origin](https://developercertificate.org/).
- [x] Check ["Allow edit from maintainers" option](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/) in pull request so that additional changes can be pushed by Gradle team
- [ ] Provide integration tests (under `<subproject>/src/integTest`) to verify changes from a user perspective
- [x] Provide unit tests (under `<subproject>/src/test`) to verify logic
- [ ] Update User Guide, DSL Reference, and Javadoc for public-facing changes
- [x] Ensure that tests pass locally: `./gradlew <changed-subproject>:check`

### Gradle Core Team Checklist
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation
- [ ] Recognize contributor in release notes

### CODE REVIEW

File: subprojects/ide/src/main/java/org/gradle/plugins/ide/internal/tooling/EclipseModelBuilder.java [hunk 1]
```diff
@@ -196,54 +198,88 @@ private void populate(Project project) {
             }
         }
 
+        // org.gradle.plugins.ide.eclipse.model.NonRenamableProject  takes
```
Comment (oehme) [inline subprojects/ide/src/main/java/org/gradle/plugins/ide/internal/tooling/EclipseModelBuilder.java]: Please remove all these inline comments.

File: subprojects/ide/src/test/groovy/org/gradle/plugins/ide/internal/tooling/eclipse/EclipseModelBuilderTest.groovy [hunk 2]
```diff
@@ -66,6 +67,20 @@ class EclipseModelBuilderTest extends AbstractProjectBuilderSpec {
...
+        project.eclipse.project.natures = ['nature.a']
+        project.eclipse.project.file.beforeMerged { it.natures += 'nature.b' }
+        project.eclipse.project.file.whenMerged { it.natures += 'nature.c' }
+        def modelBuilder = createEclipseModelBuilder()
+
+        when:
+        def eclipseModel = modelBuilder.buildAll("org.gradle.tooling.model.eclipse.EclipseProject", project)
+
+        then:
+        println eclipseModel.projectNatures.collect { it.id } == ['nature.b', 'nature.a', 'nature.c']
```
Comment (oehme) [inline subprojects/ide/src/test/groovy/org/gradle/plugins/ide/internal/tooling/eclipse/EclipseModelBuilderTest.groovy]: Remove println

**Comment by donat** [comment]
@oehme Sure, I already assigned myself to this PR.

File: subprojects/ide/src/main/java/org/gradle/plugins/ide/internal/tooling/EclipseModelBuilder.java (hunk 1 above)
Comment (lukeu) [reply subprojects/ide/src/main/java/org/gradle/plugins/ide/internal/tooling/EclipseModelBuilder.java]: *smack head* Whoops sorry, it was very late.

Interesting observation though: "Project" and "EclipseProject" are very overloaded terms around these parts :-) you can see I was struggling to remember which was which, esp. when variable names often clash with different classes too...

File: subprojects/ide/src/test/groovy/org/gradle/plugins/ide/internal/tooling/eclipse/EclipseModelBuilderTest.groovy (hunk 2 above)
Comment (lukeu) [reply subprojects/ide/src/test/groovy/org/gradle/plugins/ide/internal/tooling/eclipse/EclipseModelBuilderTest.groovy]: How'd that get there?

File: subprojects/ide/src/main/java/org/gradle/plugins/ide/internal/tooling/EclipseModelBuilder.java [hunk 3]
```diff
@@ -197,14 +199,32 @@ private void populate(Project project) {
...
 
         DefaultEclipseProject eclipseProject = findEclipseProject(project);
+
         eclipseProject.setClasspath(externalDependencies);
         eclipseProject.setProjectDependencies(projectDependencies);
         eclipseProject.setSourceDirectories(sourceDirectories);
         eclipseProject.setClasspathContainers(classpathContainers);
         eclipseProject.setOutputLocation(outputLocation != null ? outputLocation : new DefaultEclipseOutputLocation("bin"));
 
+        // Create as a holder object.
```
Comment (lukeu) [inline subprojects/ide/src/main/java/org/gradle/plugins/ide/internal/tooling/EclipseModelBuilder.java]: Was this included in the "remove all the inline comments" request? I see comments are fairly sparse around here (self-documenting code FTW) but I think that anything that might be "surprising" should be explained.

Comment (oehme) [reply subprojects/ide/src/main/java/org/gradle/plugins/ide/internal/tooling/EclipseModelBuilder.java]: Yes, this was included as well, as it only repeats what the code is saying.

File: subprojects/ide/src/main/java/org/gradle/plugins/ide/internal/tooling/EclipseModelBuilder.java [hunk 4]
```diff
@@ -197,53 +199,72 @@ private void populate(Project project) {
...
         eclipseProject.setClasspath(externalDependencies);
         eclipseProject.setProjectDependencies(projectDependencies);
         eclipseProject.setSourceDirectories(sourceDirectories);
+        eclipseProject.setClasspathContainers(classpathContainers);
+        eclipseProject.setOutputLocation(outputLocation != null ? outputLocation : new DefaultEclipseOutputLocation("bin"));
 
-        List<DefaultEclipseLinkedResource> linkedResources = new LinkedList<DefaultEclipseLinkedResource>();
-        for (Link r : eclipseModel.getProject().getLinkedResources()) {
-            linkedResources.add(new DefaultEclipseLinkedResource(r.getName(), r.getType(), r.getLocation(), r.getLocationUri()));
+        // The new Project means 'beforeMerged' won't get to see any previous state from the .project file.
```
Comment (lukeu) [inline subprojects/ide/src/main/java/org/gradle/plugins/ide/internal/tooling/EclipseModelBuilder.java]: Yeah, it was a bit wordy. Can I interest you in this one final line? ;-) Fine if not.

**Comment by donat** [comment]
@lukeu any update here?

**Comment by lukeu** [comment]
Ah sorry, I was waiting to hear back from you on this one. The WIP was just a failsafe to make sure I heard back from some buildship folk before someone else merged it, so If you think it's good to go I can remove that. I'll rebase it and do that. Anything else to do?

File: subprojects/ide/src/main/java/org/gradle/plugins/ide/internal/tooling/EclipseModelBuilder.java (hunk 4 above)
Comment (donat) [reply subprojects/ide/src/main/java/org/gradle/plugins/ide/internal/tooling/EclipseModelBuilder.java]: ❌ IMO the comment adds more confusion than clarification. Let's get rid of it.

File: subprojects/ide/src/test/groovy/org/gradle/plugins/ide/internal/tooling/eclipse/EclipseModelBuilderTest.groovy [hunk 5]
```diff
@@ -66,6 +67,20 @@ class EclipseModelBuilderTest extends AbstractProjectBuilderSpec {
         eclipseModel.projectNatures.collect { it.id } == ['nature.a', 'nature.b']
     }
 
+    def "can modify natures before and after merging"() {
```
Comment (donat) [inline subprojects/ide/src/test/groovy/org/gradle/plugins/ide/internal/tooling/eclipse/EclipseModelBuilderTest.groovy]: ❌ Please add test coverage for all attributes that can be adjusted in in `project.file.beforeMerged` and `project.file.whenMerged`: 
- natures
- build commands
- linked resources

File: subprojects/ide/src/test/groovy/org/gradle/plugins/ide/internal/tooling/eclipse/EclipseModelBuilderTest.groovy [hunk 6]
```diff
@@ -66,6 +67,20 @@ class EclipseModelBuilderTest extends AbstractProjectBuilderSpec {
...
+        project.eclipse.project.file.whenMerged { it.natures += 'nature.c' }
+        def modelBuilder = createEclipseModelBuilder()
+
+        when:
+        def eclipseModel = modelBuilder.buildAll("org.gradle.tooling.model.eclipse.EclipseProject", project)
+
+        then:
+        eclipseModel.projectNatures.collect { it.id } == ['nature.b', 'nature.a', 'nature.c']
+    }
+
```
Comment (donat) [inline subprojects/ide/src/test/groovy/org/gradle/plugins/ide/internal/tooling/eclipse/EclipseModelBuilderTest.groovy]: ❌ Please add Tooling API test coverage. You'll find a very similar test [here](https://github.com/gradle/gradle/blob/a0307889bed8a9ce58e1faf005a86ad86e1fb99e/subprojects/ide/src/crossVersionTest/groovy/org/gradle/plugins/ide/tooling/r30/ToolingApiEclipseModelClasspathContainerCrossVersionSpec.groovy).

Comment (donat) [inline subprojects/ide/src/test/groovy/org/gradle/plugins/ide/internal/tooling/eclipse/EclipseModelBuilderTest.groovy]: ❌This change can break the IDE setup if somebody first used `eclipse` Gradle plugin then migrated to Buildship and left some code in`file.beforeMerged`/`file.whenMerged`. I'd suggest explaining the changes in the [release notes](https://github.com/gradle/gradle/blob/master/subprojects/docs/src/docs/release/notes.md) as a potential breaking change.

**Comment by lukeu** [comment]
Thanks @donat, I'll get on to that over the weekend / early next week. (Short day for me today.)

**Comment by lukeu** [comment]
Update: unit tests updated, but I've been struggling a bit with the Tooling API test. I had trouble getting this to compile in either Eclipse or IDEA (which I downloaded out of desperation). But gave up.

Anyway, I can fumble along editing it as text & compiling on the CL

What would help is a command or option to speed up testing. Do you know how I could either:
  * invoke just the single test I'm editing: `crossVersionTest/groovy/org/gradle/plugins/ide/tooling/r49/` `ToolingApiEclipseModelProjectCrossVersionSpec`
  * increase the number of tests that `./gradlew :ide:crossVT` runs in parallel (it runs 4, but I think it could happily do 8-10)

BTW is the name & directory for that test OK?

**Comment by lukeu** [comment]
(just for clarification, I won't push out any changes until I've responded to all your comments)

**Comment by lukeu** [comment]
> What would help is a command or option to speed up testing

I'm good now, up and running with IDEA. (The secret was to use `./gradlew idea`, but then stop clicking the "Refresh Gradle" hyperlink it suggested.)

**Comment by lukeu** [comment]
Right, I think I've finally cracked this nut. Please see the changes I've now rebased & force-pushed.

**Comment by donat** [comment]
The PR build failed due to [Maven Central TLS](https://github.com/gradle/gradle/issues/5740) problem. I've pushed the changes to a new branch and executed the `Branch Accept` pipeline manually: https://builds.gradle.org/viewQueued.html?itemId=13713603

**Comment by donat** [comment]
@lukeu Thank you for the contribution!

**Comment by lukeu** [comment]
you're welcome :-)
